### PR TITLE
8277762: Allow configuration of HOTSPOT_BUILD_USER

### PR DESCRIPTION
**Pull request by andrew-m-leonard**
To better allow "reproducible builds", a new configure parameter is added to set the USERNAME build variable, rather than always using the current user:
    --with-build-user=<username> 
HOTSPOT_BUILD_USER is then set to a reproducible USERNAME if required.

Signed-off-by: Andrew Leonard <anleonar@redhat.com>

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277762](https://bugs.openjdk.java.net/browse/JDK-8277762): Allow configuration of HOTSPOT_BUILD_USER

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6542/head:pull/6542` \
`$ git checkout pull/6542`

Update a local copy of the PR: \
`$ git checkout pull/6542` \
`$ git pull https://git.openjdk.java.net/jdk pull/6542/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6542`

View PR using the GUI difftool: \
`$ git pr show -t 6542`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6542.diff">https://git.openjdk.java.net/jdk/pull/6542.diff</a>

</details>

### CODE REVIEW

File: make/autoconf/basic.m4 [hunk 1]
```diff
@@ -96,8 +96,14 @@ AC_DEFUN_ONCE([BASIC_SETUP_PATHS],
   AUTOCONF_DIR=$TOPDIR/make/autoconf
 
   # Setup username (for use in adhoc version strings etc)
-  # Outer [ ] to quote m4.
-  [ USERNAME=`$ECHO "$USER" | $TR -d -c '[a-z][A-Z][0-9]'` ]
+  AC_ARG_WITH([build-user], [AS_HELP_STRING([--with-build-user],
```
Comment (erikj79) [inline make/autoconf/basic.m4:99]: This variable is currently initiated in a rather weird location given what it's used for. As it used to be a one liner, it didn't seem worth the effort to move it to a more suitable location, but when expanding it with a --with flag I think we should also move this to jdk-version.m4, which is the only place where it's currently used. I think it makes sense to treat this as part of the version variables as that's what it's used for.

Comment (andrew-m-leonard) [reply make/autoconf/basic.m4:99]: it's also used in make/hotspot/lib/CompileJvm.gmk, but agree moving to jdk-versions.m4 makes sense

Comment (erikj79) [reply make/autoconf/basic.m4:99]: Right, I meant the only place it's used within configure.

Comment (andrew-m-leonard) [reply make/autoconf/basic.m4]: done, moved

**Comment by andrew-m-leonard** [comment]
@erikj79 ready for review again please
thanks

**Review by erikj79 — APPROVED** [review]
Looks good!

**Pull request closed without merging**